### PR TITLE
Transform: Ticket - Data Normalization + Feature Engineering

### PR DESCRIPTION
**Pull request by Sampai28**
### Motivation
Resolves: #4

### Implemented Changes
- Integrated real embeddings using `EmbeddingService` from #8 (all-MiniLM-L6-v2)
- Integrated technical keyword extraction using `KeywordExtractor` from #11
- Replaced stub embeddings (768 zeros) with real 384-dimensional semantic vectors
- Replaced TF-IDF keywords with actual technical skills (terraform, kubernetes, aws, etc.)
- Updated `embedding_model` field from "stub-v1" to "all-MiniLM-L6-v2"
- **[NEW]** Implemented smart code truncation strategy (keeps 15 lines full, 5+5 for medium, 10+10 for large blocks)
- **[NEW]** Optimized keyword extraction with compiled regex pattern (10-50x speedup)
- **[NEW]** Added GPU support for embedding generation (5-10x faster when available)
- **[NEW]** Added progress bars (tqdm) for all transformation steps
- **[NEW]** Changed assignee default from "unknown" to pd.NA for proper null handling
- **[NEW]** Consolidated transform code from apps/transform to apps/training/training/etl/transform
- **[NEW]** Successfully processed 61,271 tickets (up from 900) with real ML features from all 3 repos
- Fixed linting issues in transformation pipeline files

## How has this been tested?
- Ran complete transformation pipeline on 61,271 tickets from comprehensive dataset
- Verified embeddings are real 384-dimensional floating-point values (not zeros)
- Verified keywords extract actual tech skills (terraform, kubernetes, aws, docker, etc.)
- Verified smart truncation preserves code context appropriately
- **[NEW]** Added transformation pipeline tests (test_transform_pipeline.py)
- All checks pass: `just check` (linting, formatting, type checking, tests)
- **[NEW]** GPU acceleration verified working (torch.cuda.is_available() = True)
- Output validated in `data/github_issues/tickets_transformed_improved.jsonl`

## Screenshots
N/A

## Checklist:
- [x] I followed code style of the project.
- [x] I've updated all documentation accordingly.
- [x] I've added tests to cover my changes.
- [x] I've ensured all tests pass (no regressions).

### CODE REVIEW

File: apps/transform/run_transform.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  print(f"Loaded {len(df)} tickets")
+
+  # Ensure required fields exist
+  df["title"] = df.get("title", "")
+  df["body"] = df.get("body", "")
+  df["createdAt"] = df.get("createdAt")
+  df["closedAt"] = df.get("closedAt")
+
+  if "assignee" not in df.columns:
+    df["assignee"] = "unknown"
```
Comment (ALearningCurve) [inline apps/transform/run_transform.py]: Out of curiosity - what is the reason for filling default value with "unknown" and "mid"?

File: pyproject.toml [hunk 2]
```diff
@@ -5,6 +5,10 @@ description = "Machine Learning project monorepo"
 readme = "README.md"
 requires-python = ">=3.12"
 dependencies = [
+    "pandas>=2.2.0",
```
Comment (ALearningCurve) [inline pyproject.toml]: Lets prefer to add the dependencies under the lib/app that needs them rather than at the root.

You can do something like `uv add --package mlcore numpy scikit-learn  tqdm pandas`

**Review by ALearningCurve — APPROVED** [review]

File: apps/transform/run_transform.py (hunk 1 above)
Comment (Sampai28) [reply apps/transform/run_transform.py]: In `apps\training\training\etl\ingest\scrape_github_issues.py`, we already filter out tickets without assignees during the scraping process, so we shouldn't have missing assignees in the raw data. 
The 'unknown' default in run_transform.py is defensive programming - a safety net for edge cases like potential data corruption. 
For 'mid' seniority, it's a placeholder since we don't have actual engineer seniority data yet. Once we integrate real engineer profiles, this default won't be needed.
I can remove these if you think they're unnecessary

File: apps/transform/temporal_features.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      created_at: ISO datetime string when ticket was created
+      closed_at: ISO datetime string when ticket was closed
+
+  Returns:
+      Business hours or None if closed_at is missing
+  """
+  if closed_at is None or pd.isna(closed_at):
+    return None
+
+  if created_at is None or pd.isna(created_at):
```
[the remark this answers is not in the document]
Comment (Sampai28) [reply apps/transform/temporal_features.py]: Great catch! I checked our current data and confirmed we don't have `assigned_at `timestamps yet. The scraper only captures `created_at `and `closed_at`. I ran a test and verified GitHub does provide assignment timestamps through the events API.
I agree we should track `assigned_at` to `closed_at `for more accurate work duration. For this PR, I'll keep the current implementation. I'll create a follow-up ticket to update the scraper to capture assignment timestamps from GitHub events and modify the temporal features accordingly.
Sound good?

File: pyproject.toml (hunk 2 above)
Comment (Sampai28) [reply pyproject.toml]: Good point! I'll move the dependencies to their specific packages. Going forward, I'll add dependencies directly to the right `pyproject.toml` instead of the root.

File: apps/transform/run_transform.py (hunk 1 above)
Comment (Sampai28) [reply apps/transform/run_transform.py]: Got it! So we need two separate datasets:

- Training data (current scraper): Closed tickets with assignees - used for ML model training and evaluation
- Active tickets (new scraper): All open/unassigned tickets - used for the swimlane/Jira view

Should I create a separate scraper function for open/unassigned tickets, or modify the current scraper to output two files (tickets_training.csv for ML and tickets_active.csv for the UI)? 
Since I'm anyways creating a follow-up ticket to add `assigned_at `timestamps, I can include this change there - updating the scraper to produce both datasets in the same update. 

What's your preference?

**Review by ALearningCurve — CHANGES_REQUESTED** [review]
Lets not make a new app, this transform code should be part of the pipeline code which lives in the `training` app.

Lets also add tests please and resolve this [last comment](https://github.com/ALearningCurve/ticket-forge/pull/48#discussion_r2830929736). Thanks!
